dwave/optimization/model.pyx

```diff
@@ -845,53 +846,72 @@ cdef class Model:
-        """
-        # Todo: adapt to use iter_symbols()
-        # This whole method will need a re-write, it currently only works with gcc
-        # but it is useful for development
+            >>> one = model.constant(1)
+            >>> two = model.constant(2)
+            >>> i = model.integer()
+            >>> model.minimize(two * i - one)
+            >>> G = model.to_networkx()
 
-        import re
-        import networkx
+            One advantage of converting to NetworkX is the wide availability
+            of drawing tools. See NetworkX's
+            `drawing <https://networkx.org/documentation/stable/reference/drawing.html>`_
+            documentation.
 
-        G = networkx.DiGraph()
+            This example uses a `DAGVIZ <https://wimyedema.github.io/dagviz/>`_ to
+            draw the NetworkX graph created in the example above.
 
-        cdef cppNode* ptr
-        for i in range(self._graph.num_nodes()):
-            ptr = self._graph.nodes()[i].get()
+            >>> import dagviz                      # doctest: +SKIP
+            >>> r = dagviz.render_svg(G)           # doctest: +SKIP
+            >>> with open("model.svg", "w") as f:  # doctest: +SKIP
+            ...     f.write(r)
 
-            # this regex is compiler specific! Don't do this for the general case
-            match = re.search("\d+([a-zA-z]+Node)", str(typeid(deref(ptr)).name()))
-            if not match:
-                raise ValueError
+            This creates the following image:
 
-            u = (match[1], <long>(ptr))
+            .. figure:: /_images/to_networkx_example.svg
+               :width: 500 px
+               :name: to-networkx-example
+               :alt: Image of NetworkX Directed Graph
 
-            G.add_node(u)
+        """
+        import networkx
 
-            for j in range(ptr.predecessors().size()):
-                pptr = ptr.predecessors()[j]
+        G = networkx.MultiDiGraph()
 
-                match = re.search("\d+([a-zA-z]+Node)", str(typeid(deref(pptr)).name()))
-                if not match:
-                    raise ValueError
+        # Add the symbols, in order if we happen to be topologically sorted
+        G.add_nodes_from(repr(symbol) for symbol in self.iter_symbols())
 
-                v = (match[1], <long>(pptr))
+        # Sanity check. If several nodes map to the same symbol repr we'll see
+        # too few nodes in the graph
+        if len(G) != self.num_symbols():
+            raise RuntimeError("symbol repr() is not unique to the underlying node")
 
-                G.add_edge(v, u)
+        # Now add the edges
+        for symbol in self.iter_symbols():
+            for successor in symbol.iter_successors():
+                G.add_edge(repr(symbol), repr(successor))
+
+        # Sanity check again. If the repr of symbols isn't unique to the underlying
+        # node then we'll see too many nodes in the graph here
+        if len(G) != self.num_symbols():
```

Comment:
This should not happen, right? Do you want to add some debugging info on the offending symbol for if it does and the user contacts Support?

Reply:
This is for maintainers, `RuntimeError` is not user facing unless something has gone terribly wrong.

Reply:
Can a successor ever _not_ be in `iter_symbols`? Otherwise I don't see the use for this check either.

Reply:
If the reprs are not unique, then there will be too many symbols here. Because `iter_sucessors()` makes new python objects each time. So without c706ae8 they would appear as extra nodes. This is a hedge on future changes to `__repr__()`.

Reply:
Ah, yeah, that makes sense.